ksonnet-gen/kubespec/parsing.go

```diff
@@ -111,14 +77,86 @@ func ParseDefinitionName(dn DefinitionName) *ParsedDefinitionName {
 			Codebase:    codebase,
 			Group:       nil,
 			Version:     nil,
-			Kind:        split[5],
+			Kind:        ObjectKind(split[5]),
 		}
 	}
 
-	log.Fatalf("Failed to parse definition name '%s'", string(dn))
+	log.Fatalf("Failed to parse definition name '%s'", string(*dn))
 	return nil
 }
 
+// Name parses a `DefinitionName` from an `ObjectRef`. `ObjectRef`s
+// that refer to a definition contain two parts: (1) a special prefix,
+// and (2) a `DefinitionName`, so this function simply strips the
+// prefix off.
+func (or *ObjectRef) Name() *DefinitionName {
+	defn := "#/definitions/"
+	ref := string(*or)
+	if !strings.HasPrefix(ref, defn) {
+		log.Fatalln(ref)
+	}
+	name := DefinitionName(strings.TrimPrefix(ref, defn))
+	return &name
+}
+
+//-----------------------------------------------------------------------------
+// Parsed definition name.
+//-----------------------------------------------------------------------------
+
+// Package represents the type of the definition, either `APIs`, which
+// have API groups (e.g., extensions, apps, meta, and so on), or
+// `Core`, which does not.
+type Package int
+
+const (
+	// Core is a package that contains the Kubernetes Core objects.
+	Core Package = iota
+
+	// APIs is a set of non-core packages grouped loosely by semantic
+	// functionality (e.g., apps, extensions, and so on).
+	APIs
+
+	//
+	// Internal packages.
+	//
+
+	// Util is a package that contains utilities used for both testing
+	// and running Kubernetes.
+	Util
+
+	// Runtime is a package that contains various utilities used in the
+	// Kubernetes runtime.
+	Runtime
+
+	// Version is a package that supplies version information collected
+	// at build time.
+	Version
+)
+
+// ParsedDefinitionName is a parsed version of a fully-qualified
+// OpenAPI spec name. For example,
+// `io.k8s.kubernetes.pkg.api.v1.Container` would parse into an
+// instance of the struct below.
+type ParsedDefinitionName struct {
+	PackageType Package
+	Codebase    string
+	Group       *GroupName
```

Comment:
Ah -- makes sense.  That is a standard pattern to have a pointer to string if it is optional. Another option is to just have a blank string.  In this case blank and null can be equiv and it may make other code smoother.  Either way is cool.